README.md
<h1 align="center">
  <img src="https://img.ideal-postcodes.co.uk/UK%20Clear%20Addressing%20Logo@3x.png" alt="UK Clear Addressing">
</h1>

> Correctly parse and format UK Addresses in Royal Mail's Postcode Address File

![CI](https://github.com/ideal-postcodes/uk-clear-addressing/workflows/CI/badge.svg)
![Release](https://github.com/ideal-postcodes/uk-clear-addressing/workflows/Release/badge.svg)
[![Try uk-clear-addressing on RunKit](https://badge.runkitcdn.com/uk-clear-addressing.svg)](https://npm.runkit.com/uk-clear-addressing)

Parses Postcode Address File records into correctly formatted address recognised by Royal Mail according to its Clear Addressing Guidelines.

Produces consistent address lines, a post town line and a postcode line.

## Features

![Correct Addressing](https://img.ideal-postcodes.co.uk/correct_address.gif)

- Correctly format UK addresses using Royal Mail's Postcode Address File
- Produces 3 address lines suitable for general use (i.e. mailing)
- Produces a `premise`, `number` and `unit` attribute, which is a sensible almagamation of `building_name`, `sub_building_name` and `building_number`
- Address sorting function
- Extensive test suite, including documented and newly discovered corner cases

## Links

- [API Documentation](https://uk-clear-addressing.ideal-postcodes.dev)
- [More information on Postcode Address File data attributes](https://ideal-postcodes.co.uk/documentation/paf-data)
- [PAF Programmer's Guide](https://js.ideal-postcodes.co.uk/guide.pdf)
- [Try uk-clear-addressing on RunKit](https://npm.runkit.com/uk-clear-addressing)
- [Consume this library as a HTTP Service with PAF API](https://github.com/ideal-postcodes/paf-api)

## Getting Started

### Installation

```bash
npm install uk-clear-addressing
```

### Formatting Addresses

#### Extract formatted address lines

Use the [Address](https://uk-clear-addressing.ideal-postcodes.dev/classes/index.address.html) class to parse a [PAF Record](https://uk-clear-addressing.ideal-postcodes.dev/interfaces/pafrecord.html)

[Formatted address lines](https://uk-clear-addressing.ideal-postcodes.dev/interfaces/formattedaddress.html) can be extracted using instance accessors like `line_1`, `line_2`, `line_3`, `premise`, `number` and `unit`

```javascript
const { Address } = require("uk-clear-addressing");

const pafRecord = {
  postcode: "WS11 5SB",
  post_town: "CANNOCK",
  thoroughfare: "Pye Green Road",
  building_name: "Flower House 189A",
  organisation_name: "S D Alcott Florists",
};

const {
  line_1,
  line_2,
  line_3,
  premise,
  number,
  unit,
  post_town,
  postcode
} = new Address(pafRecord);
```

#### Extract a formatted address object

Alternatively, extract a [formatted address object](https://uk-clear-addressing.ideal-postcodes.dev/interfaces/formattedaddress.html) using the [`formattedAddress`](https://uk-clear-addressing.ideal-postcodes.dev/classes/index.address.html#formattedaddress) method.

```javascript
const { Address } = require('uk-clear-addressing');

const address = new Address({
  postcode: "WS11 5SB",
  post_town: "CANNOCK",
  thoroughfare: "Pye Green Road",
  building_name: "Flower House 189A",
  organisation_name: 'S D Alcott Florists',
});

console.log(address.formattedAddress());
//  {
//    postcode: "WS11 5SB",
//    post_town: "CANNOCK",
//    line_1: "S D Alcott Florists",
//    line_2: "Flower House",
//    line_3: "189a Pye Green Road",
//    number: "189a",
//    unit: "",
//    premise: "Flower House, 189a"
//  }
```

### Non-PAF Computed Attributes

`building_number`, `building_name` and `sub_building_name` represent raw data from Royal Mail's PAF and can be difficult to parse if you are unaware of how the PAF premise fields work together. For this reason, `uk-clear-addressing` also provides computed attributes, which attempt to reasonable capture common premise concepts like building number and unit.

Due to years of accumulated complexity, few assumptions should be made of ways UK addresses can be described in terms of number, name and unit. It's recommeded to simply consume the address lines (`line_1`, `line_2`, `line_3`) outlined in [Royal Mail's Programmers Guide](https://js.ideal-postcodes.co.uk/guide.pdf). Building number, unit and PAF premise attributes should be stored as useful extras.

Some examples of this complexity include:

- Alphanumeric identifiers like `10A` are not necessarily the building number. In some instances they represent a premise with building number `10` and sub building idenitifer `A`
- The `building_number` of an address may be split across `building_name`, `building_number` and `sub_building_name`
- Some addresses may be identified as a unit `sub_building_name` but no building number or name
- Some apparent building numbers cannot be separated from the building name. For instance, `1 Ashgate Rise, Raw Gap, HG50HZ`, `1 Ashgate Rise` appears in `building_name`. The numeric element should not be separated into `number` because the property is numbered into the building and not the thoroughfare

Further keep in mind that UK addresses are manually added and updated by thousands of postal workers, thousands of times a day across this UK. Although Royal Mail strives to keep address formatting in accordance with its internal guidance, it is possible for building numbers and units may be incorrectly designated for a time.

#### `premise` attribute

The `premise` attribute is designed to capture *only* the premise specific elements of an address (i.e. no thoroughfare, localities, etc). It attempts to sensibly combine `building_number`, `building_name` and `sub_building_name`.

`premise` omits organisation, department and PO boxes.

Examples include:

```text
10B Barry Tower, 13
Flat 1-3, 10
Flat 3, Nelson House, 2
Mansion House
Suite 1-3
```

#### `number` attribute

The `number` attribute attempts to capture the building number element of an address, in other words how a premise is numbered into a thoroughfare. For simple use cases, this attribte may be prefered over Royal Mail's `building_number` attribute.

Royal Mail's `building_number` field is not suited to this task as the upstream data schema only allows this to be integers. Therefore, mixed numbers (like e.g. `1A` and `1-3`) and ordinal building identifiers (like `A`,`B`,`C`, etc) are not captured in `building_number`.

Examples include:

```text
10
A
1-3
10A
```

#### `unit` attribute

The `unit` attribute attempts to capture the "sub building" element of a building with its own number, name or both. For simple use cases, this attribute may be preferred to Royal Mail's `sub_building_name` field.

Examples include:

```text
A
Flat 1
Basement Flat
Caretakers Flat
```

### Sorting Addresses

[`Address.sort`](https://uk-clear-addressing.ideal-postcodes.dev/modules/sort.html#sort-1) implements a comparison function, which allows you to compare [`Address`](https://uk-clear-addressing.ideal-postcodes.dev/classes/index.address.html) instances. This can readily be passed into `Array.prototype.sort`

```javascript
const addresses = await query("SELECT * FROM postcode_address_file LIMIT 10");

addresses
  .map(address => new Address(address)) // Instantiate an `Address` instances
  .sort(Address.sort)                   // Now sort
  // Print an example to console
  .forEach(address => console.log(address.line_1));
  // "190 Elm Road"
  // "190a Elm Road"
  // "191 Elm Road"
  // "191a Elm Road"
  // "192 Elm Road"
  // "193 Elm Road"
  // "193a Elm Road"
  // "197 Elm Road"
  // "197a Elm Road"
  // "199 Elm Road"
```

## Testing

Many of the regular and edge cases are documented in the test. To run the test suite:

```bash
npm test
```

If you find an edge case, please feel free to make a pull request. However be sure to include a test which documents the specific case being handled.

## Parameters

Below is a list of address fragments. For the address to be properly formatted, you need to pass in all the address fragments available to you.

### Premises Elements

- Sub Building Name (e.g. ‘Flat 1’)
- Building Name (e.g. ‘Rose Cottage’)
- Building Number (e.g. ‘22’)
- Organisation Name (e.g. ‘Cath’s Cakes’)
- PO Box number

### Thoroughfare Elements

- Dependant Thoroughfare Name (e.g. ‘Cheshunt’)
- Dependant Thoroughfare Descriptor (e.g. ‘Mews’ or ‘Court’)
- Thoroughfare Name (e.g. ‘Cypress’)
- Thoroughfare Descriptor (e.g. ‘Road’ or ‘Street’)

### Locality Elements

- Double Dependant Locality (e.g. ‘Tyre Industrial Estate’)
- Dependant Locality (e.g. ‘Blantyre’)
- Post Town (e.g. ‘GLASGOW’)

## Licence

MIT


CHANGELOG.md
## [4.2.1](https://github.com/ideal-postcodes/uk-clear-addressing/compare/4.2.0...4.2.1) (2023-12-14)


### Bug Fixes

* **Acronyms:** Building name check ([f43a741](https://github.com/ideal-postcodes/uk-clear-addressing/commit/f43a74157102c5a385f49eedb90427acf0d23544))

# [4.2.0](https://github.com/ideal-postcodes/uk-clear-addressing/compare/4.1.1...4.2.0) (2021-12-31)


### Features

* **Sort:** Export pure sort scoring method ([2e6b9de](https://github.com/ideal-postcodes/uk-clear-addressing/commit/2e6b9def89c4079df5835c751a9291f33003fa41))

## [4.1.1](https://github.com/ideal-postcodes/uk-clear-addressing/compare/4.1.0...4.1.1) (2021-11-17)


### Bug Fixes

* **Sub + Building Name:** Detect address range ([cbffe30](https://github.com/ideal-postcodes/uk-clear-addressing/commit/cbffe306d3d9fff28fac958acebaaa480acc85f4))

# [4.1.0](https://github.com/ideal-postcodes/uk-clear-addressing/compare/4.0.0...4.1.0) (2021-05-17)


### Features

* **Sort:** Export sorting attributes ([f0625f4](https://github.com/ideal-postcodes/uk-clear-addressing/commit/f0625f415b24175b5f207dc985c740ff167daa9d))

# [4.0.0](https://github.com/ideal-postcodes/uk-clear-addressing/compare/3.0.0...4.0.0) (2021-03-05)


### Bug Fixes

* **Release:** Trigger release ([764d1e4](https://github.com/ideal-postcodes/uk-clear-addressing/commit/764d1e4c33c505ac2445e89ba8ba78939cdc72b1))


### Documentation

* **Premise Attributes:** Document number and unit ([a363dc9](https://github.com/ideal-postcodes/uk-clear-addressing/commit/a363dc9f062849b254536066917f568f3b1f39d5))


### Features

* **Building Numbers:** Experimental attempt at returning actual ([0f653e9](https://github.com/ideal-postcodes/uk-clear-addressing/commit/0f653e99af741109b7915bb30071293e3252f411))
* **Building Units:** Experimental attempt at returning a building unit ([e72d7c4](https://github.com/ideal-postcodes/uk-clear-addressing/commit/e72d7c4945abc85006708686dd724355061beb10))
* **ES Module:** Add ESM build ([e6b5632](https://github.com/ideal-postcodes/uk-clear-addressing/commit/e6b5632f11163bec68943fbe6f1beced2fdd6a07))


### BREAKING CHANGES

* **Premise Attributes:** - formattedADdress now returns `unit` and `number`
- FormattedPremise interface includes `unit` and `number`

# [3.0.0](https://github.com/ideal-postcodes/uk-clear-addressing/compare/2.2.2...3.0.0) (2020-12-10)


### Bug Fixes

* **Semantic Release:** Automate npm release ([06ff64f](https://github.com/ideal-postcodes/uk-clear-addressing/commit/06ff64f958166adbe2fb43c2a94c44c04204f301))


### chore

* **Node:** Drop explicit support for Node 8 ([195cc59](https://github.com/ideal-postcodes/uk-clear-addressing/commit/195cc59203c4001ac4b44d0e2232c80b9a7b9077))
* **Node:** Drop support for node 8, add 14 ([e69e64d](https://github.com/ideal-postcodes/uk-clear-addressing/commit/e69e64d130225c654ef0fcd5e4c47f219935db96))


### BREAKING CHANGES

* **Node:** Node 8 no longer supported
* **Node:** Node 8 no longer forms part of CI testing

# Changelog

## [2.2.2] 2019-04-24

- Update dependencies
- Update documentation
  - Reference PAF API
  - Reference typedocs

## [2.2.1] 2019-04-02

- Fix: Handle building names that contain range on rule 6

## [2.2.0] 2019-03-11

- Formatting updates with guidance and examples outlined in the latest PAF Programmers Guide (Edition 7, Version 6):
  - Where a building name exception exists in rule 3, these names are no longer lower cased as documented in Table 27a. They now adhere to the (conflicting) example defined in Table 22
  - Building ranges which contain specific prefixes (e.g. Back of, Stalls, Maisonette) are no longer broken up across lines
- Added `line_1`, `line_2`, `line_3` and `premise` accessor methods on `Address` instances
- Standardise project with prettier and @cablanchard/tslint
- Use stardardised compile options from @cablanchard/tsconfig
- Added benchmark suite

## [2.1.2] 2019-02-12

- Add runkit example
- Use whitelist to generate npm package

## [2.1.1] 2018-07-02

- Fix: Drop nyc_output from npm package

## [2.1.0] 2018-06-28

- Project reorganisation
	- Typescript files omitted from npm package
	- Compiled js and typing files omitted from project repository
	- Moved compiled files to `dist/`
	- Dropped unused assets from npm module
- Fix: typings for `Address.sort` (previously returned inappropriate type)
- Fix: address sort comparator
- Fix: formatting for corner case on rule6

## [2.0.0] 2018-04-05

### Breaking Changes

- Node.js 6.x and above supported only
- Premise elements are no longer capitalised by default. PAF data should be appropriately capitalised
- Address class must be extracted from the library. E.g.

```
// New method of retrieving class

const { Address } = require("uk-clear-addressing");

```

### Other Changes

- `Address.sort` implements a sorting method which allows two `address` objects to be compared
- This project has been ported to typescript. Typings now available

## [0.1.2] 2014-12-05
- formattedAddress() now returns a premise attribute. This is a computed attribute to isolate the premise name taking into account all available information.

## [0.1.1] 2014-12-05
- Added fallback for 4 line addresses (very rare edge case): When an address spills over 3 lines, the remaining lines (lines 4 or more) are appended to line 3 and separated by commas
- Added fallback for address with only a sub_building_name (very rare edge case): See `weirdRule` in index.js
- Performance Improvement: Result of formattedAddress is cached
- Performance Improvement: Removal of string dependency


src/index.ts
import {
  EmptyString,
  AddressRecord,
  RawAddress,
  AddressJSON,
  FormattedAddress,
} from "./types";
import { sort, sortingElems } from "./sort";
import { formatter } from "./rules";
import { extract, extractFloat, extractInteger } from "./utils";

/**
 * The `Address` class is designed to wrap a Postcode Address File (PAF)
 * record and provide utility methods to make this data more readily
 * consumable for humans
 *
 * `Address` requires an object of type `AddressRecord` to be instantiated
 * which is comprised of fields readily found on PAF
 *
 * `Address` can be used to
 * - Compute address lines based on premise and locality information
 * - Sensibly compute a `premise` label based on (sub) building name and number
 * - Sensibly compute `number` and `unit` labels based on (sub) building name and number
 * - Sensibly sort an array of addresses
 *
 * @example
 * ```typescript
 * // Formatting an address
 *
 * const address = new Address({
 * 	postcode: "WS11 5SB",
 * 	post_town: "CANNOCK",
 * 	dependant_locality: "",
 * 	double_dependant_locality: "",
 * 	thoroughfare: "Pye Green Road",
 * 	building_number: "",
 * 	building_name: "Flower House 189A",
 * 	sub_building_name: "",
 * 	dependant_thoroughfare: "",
 * 	organisation_name: 'S D Alcott Florists',
 * });
 *
 * console.log(address.formattedAddress());
 *
 * //
 * //	{
 * //		postcode: 'WS11 5SB',
 * //		post_town: 'CANNOCK',
 * //		line_1: 'S D Alcott Florists',
 * //		line_2: 'Flower House',
 * //		line_3: '189a Pye Green Road',
 * //		unit: ""
 * //		number: "189a"
 * //		premise: "Flower House, 189a"
 * //	}
 * //
 * ```
 *
 * @example
 * ```typescript
 * // Formatting an address
 *
 *const addresses = await query("SELECT * FROM postcode_address_file LIMIT 10");
 *
 * addresses
 * 	.map(address => new Address(address)) // Instantiate an `Address` instances
 * 	.sort(Address.sort)  								  // Now sort
 *
 * 	// Print an example to console
 * 	.forEach(address => console.log(address.line_1));
 * 	// "190 Elm Road"
 * 	// "190a Elm Road"
 * 	// "191 Elm Road"
 * 	// "191a Elm Road"
 * 	// "192 Elm Road"
 * 	// "193 Elm Road"
 * 	// "193a Elm Road"
 * 	// "197 Elm Road"
 * 	// "197a Elm Road"
 * 	// "199 Elm Road"
 * ```
 */
export class Address {
  readonly merge_sub_and_building: boolean;
  readonly postcode_inward: string;
  readonly postcode_outward: string;

  readonly po_box: string;
  readonly postcode: string;
  readonly post_town: string;
  readonly dependant_locality: string;
  readonly double_dependant_locality: string;
  readonly thoroughfare: string;
  readonly dependant_thoroughfare: string;
  readonly building_number: string;
  readonly building_name: string;
  readonly sub_building_name: string;
  readonly department_name: string;
  readonly organisation_name: string;
  readonly postcode_type: string;
  readonly su_organisation_indicator: string;
  readonly delivery_point_suffix: string;
  readonly county: string;
  readonly traditional_county: string;
  readonly administrative_county: string;
  readonly postal_county: string;
  readonly district: string;
  readonly ward: string;
  readonly country: string;

  readonly northings: number | EmptyString;
  readonly eastings: number | EmptyString;
  readonly udprn: number | EmptyString;
  readonly umprn: number | EmptyString;

  readonly longitude: number | EmptyString;
  readonly latitude: number | EmptyString;

  public cache: FormattedAddress | null;

  constructor(data: AddressRecord) {
    this.postcode = extract(data, "postcode");
    this.post_town = extract(data, "post_town").toUpperCase();
    this.dependant_locality = extract(data, "dependant_locality");
    this.double_dependant_locality = extract(data, "double_dependant_locality");
    this.thoroughfare = extract(data, "thoroughfare");
    this.dependant_thoroughfare = extract(data, "dependant_thoroughfare");
    this.building_number = extract(data, "building_number");
    this.building_name = extract(data, "building_name");
    this.sub_building_name = extract(data, "sub_building_name");
    this.department_name = extract(data, "department_name");
    this.organisation_name = extract(data, "organisation_name");
    this.postcode_type = extract(data, "postcode_type");
    this.su_organisation_indicator = extract(data, "su_organisation_indicator");
    this.delivery_point_suffix = extract(data, "delivery_point_suffix");
    this.county = extract(data, "county");
    this.traditional_county = extract(data, "traditional_county");
    this.administrative_county = extract(data, "administrative_county");
    this.postal_county = extract(data, "postal_county");
    this.district = extract(data, "district");
    this.ward = extract(data, "ward");
    this.country = extract(data, "country");
    this.po_box = extract(data, "po_box");

    // Extract integers
    this.udprn = extractInteger(data, "udprn");
    this.umprn = extractInteger(data, "umprn");
    this.northings = extractInteger(data, "northings");
    this.eastings = extractInteger(data, "eastings");

    // Extract floats
    this.longitude = extractFloat(data, "longitude");
    this.latitude = extractFloat(data, "latitude");

    if (this.postcode) {
      this.postcode_outward = this.postcode.split(" ")[0];
      this.postcode_inward = this.postcode.split(" ")[1];
    } else {
      this.postcode_outward = "";
      this.postcode_inward = "";
    }

    this.cache = null;

    if (this.building_number === "0") {
      this.building_number = "";
      this.merge_sub_and_building = true;
    } else {
      this.merge_sub_and_building = false;
    }
  }

  get line_1(): string {
    return this.formattedAddress().line_1;
  }

  get line_2(): string {
    return this.formattedAddress().line_2;
  }

  get line_3(): string {
    return this.formattedAddress().line_3;
  }

  get premise(): string {
    return this.formattedAddress().premise;
  }

  get unit(): string {
    return this.formattedAddress().unit;
  }

  get number(): string {
    return this.formattedAddress().number;
  }

  /**
   * Returns object containing raw address attributes
   * @hidden
   */
  raw(): RawAddress {
    return {
      postcode: this.postcode,
      post_town: this.post_town,
      dependant_locality: this.dependant_locality,
      double_dependant_locality: this.double_dependant_locality,
      thoroughfare: this.thoroughfare,
      dependant_thoroughfare: this.dependant_thoroughfare,
      building_number: this.building_number,
      building_name: this.building_name,
      sub_building_name: this.sub_building_name,
      department_name: this.department_name,
      organisation_name: this.organisation_name,
      postcode_type: this.postcode_type,
      su_organisation_indicator: this.su_organisation_indicator,
      delivery_point_suffix: this.delivery_point_suffix,
      county: this.county,
      traditional_county: this.traditional_county,
      administrative_county: this.administrative_county,
      postal_county: this.postal_county,
      district: this.district,
      ward: this.ward,
      country: this.country,
      po_box: this.po_box,
      udprn: this.udprn,
      umprn: this.umprn,
      northings: this.northings,
      eastings: this.eastings,
      longitude: this.longitude,
      latitude: this.latitude,
      postcode_outward: this.postcode_outward,
      postcode_inward: this.postcode_inward,
    };
  }

  /**
   * Returns a complete JSON representation of the `Address` instance
   */
  toJSON(): AddressJSON {
    return Object.assign(this.raw(), this.formattedAddress());
  }

  /**
   * Returns an object representing an address with sensibly computed address
   * line labels according to Royal Mail's formatting rules
   */
  formattedAddress(): FormattedAddress {
    if (this.cache) return this.cache;
    this.cache = Object.assign(
      {
        post_town: this.post_town,
        postcode: this.postcode,
      },
      formatter(this)
    );
    return this.cache;
  }

  static formatPostcode(postcode: string): string {
    return postcode
      .toString()
      .toUpperCase()
      .replace(/\s/g, "")
      .replace(/%20/g, "");
  }

  /**
   * A function which allow two `Address` objects to be compared.
   * This function can be readily fed into `Array.prototype.sort`
   *
   * @example
   *
   * ```
   * addresses.sort(Address.sort)
   * ```
   */
  static sort(a: Address, b: Address): number {
    return sort(a, b);
  }
}

export { AddressRecord, sortingElems, sort };
